app/uk/gov/hmrc/tai/service/TaxCodeChangeService.scala

```diff
@@ -116,17 +116,49 @@ class TaxCodeChangeServiceImpl @Inject()(taxCodeChangeConnector: TaxCodeChangeCo
     }
   }
 
-  private def addEndDate(date: LocalDate, taxCodeRecord: TaxCodeRecord): TaxCodeRecordWithEndDate ={
+  private def addEndDate(date: LocalDate, taxCodeRecord: TaxCodeRecord): TaxCodeRecordWithEndDate = {
     TaxCodeRecordWithEndDate(
       taxCodeRecord.taxCode, taxCodeRecord.basisOfOperation, taxCodeRecord.dateOfCalculation, date,
       taxCodeRecord.employerName, taxCodeRecord.payrollNumber, taxCodeRecord.pensionIndicator, taxCodeRecord.isPrimary
     )
   }
 
+  def latestTaxCodes(nino:Nino, taxYear: TaxYear)(implicit hc: HeaderCarrier):Future[Seq[TaxCodeRecordWithEndDate]] = {
+      mostRecentTaxCodeRecordInYear(nino, taxYear)
+  }
+
   private def taxCodeHistory(nino: Nino, taxYear: TaxYear): Future[TaxCodeHistory] = {
     taxCodeChangeConnector.taxCodeHistory(nino, taxYear, taxYear.next)
   }
 
+  def mostRecentTaxCodeRecordInYear(nino: Nino, year: TaxYear): Future[Seq[TaxCodeRecordWithEndDate]] = {
+
+    taxCodeChangeConnector.taxCodeHistory(nino, year, year.next).map { taxCodeHistory =>
+
+      val groupedTaxCodeRecords: Map[String, Seq[TaxCodeRecord]] = taxCodeHistory.taxCodeRecords.groupBy(_.employerName)
+
+      val thing = groupedTaxCodeRecords.values.flatMap {
```

Comment:
The flatMap here means that we don't need the `groupBy` on line 138. This should probably be something more along the line of 

```suggestion
      taxCodeHistory.taxCodeRecords.map {
```